src/services/codeFixProvider.ts

```diff
@@ -1,40 +1,56 @@
 /* @internal */
 namespace ts {
-    export interface CodeFix {
+    export interface CodeFixRegistration {
         errorCodes: number[];
-        getCodeActions(context: CodeFixContext): CodeAction[] | undefined;
+        getCodeActions(context: CodeFixContext): CodeFix[] | undefined;
+        groupIds: string[];
+        fixAllInGroup(context: CodeFixAllContext): CodeActionAll;
     }
 
-    export interface CodeFixContext extends textChanges.TextChangesContext {
-        errorCode: number;
+    export interface CodeFixContextBase extends textChanges.TextChangesContext {
         sourceFile: SourceFile;
-        span: TextSpan;
         program: Program;
         host: LanguageServiceHost;
         cancellationToken: CancellationToken;
     }
 
+    export interface CodeFixAllContext extends CodeFixContextBase {
+        groupId: {};

```

Comment:
i think we should just make this a string, unless we have a reason not to.

Reply:
If we use string we are limited to treating it like an enum. With an object we could support any data as part of an action id, such as the name of a class if we wanted to limit fixes to just that. That *could* be done with a string, but we would have to stringify and parse the data, which is silly because we're sending this through JSON already anyway.